### DISCUSSION TITLE
How to restore original implementation and continue syping on a method?

### DISCUSSION THREAD

**Discussion by Lordfirespeed** [vitest-dev/vitest#6308] [Q&A]
Title. https://github.com/vitest-dev/vitest/pull/3386 is relevant

- `mockClear` doesn't affect implementation 
- `mockReset` restores an empty implementation (not the original)
- `mockRestore` stops spying

example:
```ts
import { it, vi, afterEach, expect } from "vitest"

const foo = {
  bar: () => "Hello, world!"
}

vi.spyOn(foo, 'bar')

afterEach(() => {
  vi.restoreAllMocks() // the problem area: `restoreAllMocks` un-spies
})

it("Should be spied", () => {
  vi.mocked(foo.bar).mockReturnValue("Placeholder")
  foo.bar()
  expect(foo.bar).toHaveBeenCalled()
})

it("Should still be spied", () => {
  foo.bar()
  expect(foo.bar).toHaveBeenCalled()
})
```

The first test succeeds, but the second test fails as `restoreAllMocks` un-spies `foo.bar`.

How can I restore original implementations after each test, without un-spying?

**Comment by sheremet-va** [comment]
If you want others to help, you should probably provide some code so it is easier to understand what you mean.

**Reply by Lordfirespeed** [reply]
Done.

**Comment by sheremet-va** [comment]
You cannot do that without losing the spy assignment, you need to re-apply the spy again. This is done to be compatible with Jest.

**Reply by Lordfirespeed** [reply]
Right, I can see that, could I open a feature request for adding an option to `mockRestore`/`restoreAllMocks` that causes those functions not to unspy ?

**Reply by sheremet-va** [reply]
I don't think it makes sense to not unspy `mockRestores` - it's in the name, it restores the original state

**Reply by Lordfirespeed** [reply]
I disagree wholeheartedly. It makes very little sense to me that `mockRestore` would unspy. It's certainly not explicitly mentioned in the docs!

Clearly someone else thought it made sense to have `mockRestore` keep spies at some point, or #3127 would never have existed. 
For example, there are some scenarios where it is not possible to re-apply the spy (e.g. module mocks). 

Just to clarify, here is what I am proposing:
```ts
import { vi } from "vitest"

const foo = {
    bar: () => "Hello, world!"
}

vi.spyOn(foo, "bar")

vi.mocked(foo.bar).mockRestore({ unspy: false }) // restores original impl without unspying
vi.mocked(foo.bar).mockRestore() // restores original `foo.bar`, unspying

vi.restoreAllMocks({ unspy: false }) // restores original impl for all mocks without unspying
vi.restoreAllMocks() // restores originals for all mocks, unspying
```
and adding an option e.g. `mockRestoreUnspyDefault` to the vitest config, `true` by default.

I don't see why you could possibly have a problem with that - the intent is very obvious. Compatibility with Jest is kept, the library becomes more flexible.

What do you think would make sense instead?

**Reply by sheremet-va** [reply]
> Clearly someone else thought it made sense to have mockRestore keep spies at some point, or https://github.com/vitest-dev/vitest/issues/3127 would never have existed.

Someone else (me) introduced a bug that he (me) fixed in #3127

I agree that it is unfortunate when we can't reapply a spy after the mock has been restored.

I think if we want to introduce this behavior, then it should be a separate method. `mockRestore` correctly removes the spy from the object because it _restores_ the original implementation.

**Reply by Lordfirespeed** [reply]
> Someone else (me) introduced a bug that he (me) fixed in https://github.com/vitest-dev/vitest/issues/3127

:joy:

> I think if we want to introduce this behavior, then it should be a separate method. mockRestore correctly removes the spy from the object because it restores the original implementation.

Okay, cool! I'd be plenty happy with that - not sure what to name such a thing, though. We already have `mockClear`, `mockReset`, `mockRestore`. Maybe `mockRevert`?

**Reply by sheremet-va** [reply]
`mockRever` sounds good to me. You can open an issue or a PR for that

**Reply by Lordfirespeed** [reply]
Opened #6378
